src/trajectory.jl

```diff
@@ -668,79 +600,75 @@ function generalised_uturn_criterion(rho, p_sharp_minus, p_sharp_plus)
 end
 
-function isterminated(h::Hamiltonian, t::BinaryTree{T}, args...) where {T<:Union{ClassicNoUTurn, GeneralisedNoUTurn}}
-    return isterminated(h, t)
-end
-
-"""
-Recursivly build a tree for a given depth `j`.
-"""
+"Recursively build a tree for a given depth `j`."
 function build_tree(
     rng::AbstractRNG,
-    nt::NUTS{S,C,I,F},
+    integrator::AbstractIntegrator,
+    tc::AbstractTerminationCriterion,
     h::Hamiltonian,
     z::PhasePoint,
-    sampler::AbstractTrajectorySampler,
+    sampler::TS,
     v::Int,
     j::Int,
     H0::AbstractFloat,
-) where {I<:AbstractIntegrator,F<:AbstractFloat,S<:AbstractTrajectorySampler,C<:AbstractTerminationCriterion}
+) where {TS}
     if j == 0
         # Base case - take one leapfrog step in the direction v.
-        z′ = step(nt.integrator, h, z, v)
+        z′ = step(integrator, h, z, v)
         H′ = energy(z′)
         ΔH = H′ - H0
         α′ = exp(min(0, -ΔH))
-        sampler′ = S(sampler, H0, z′)
-        return BinaryTree(z′, z′, C(z′), α′, 1, ΔH), sampler′, Termination(sampler′, nt, H0, H′)
+        sampler′ = TS(sampler, H0, z′)
+        return BinaryTree(z′, z′, tc, α′, 1, ΔH), sampler′, Termination(sampler′, tc, H0, H′)
     else
         # Recursion - build the left and right subtrees.
-        tree′, sampler′, termination′ = build_tree(rng, nt, h, z, sampler, v, j - 1, H0)
+        tree′, sampler′, termination′ = build_tree(rng, integrator, tc, h, z, sampler, v, j - 1, H0)
         # Expand tree if not terminated
         if !isterminated(termination′)
             # Expand left
             if v == -1
-                tree′′, sampler′′, termination′′ = build_tree(rng, nt, h, tree′.zleft, sampler, v, j - 1, H0) # left tree
+                tree′′, sampler′′, termination′′ = build_tree(rng, integrator, tc, h, tree′.zleft, sampler, v, j - 1, H0) # left tree
                 treeleft, treeright = tree′′, tree′
             # Expand right
             else
-                tree′′, sampler′′, termination′′ = build_tree(rng, nt, h, tree′.zright, sampler, v, j - 1, H0) # right tree
+                tree′′, sampler′′, termination′′ = build_tree(rng, integrator, tc, h, tree′.zright, sampler, v, j - 1, H0) # right tree
                 treeleft, treeright = tree′, tree′′
             end
             tree′ = combine(treeleft, treeright)
             sampler′ = combine(rng, sampler′, sampler′′)
-            termination′ = termination′ * termination′′ * isterminated(h, tree′, treeleft, treeright)
+            termination′ = termination′ * termination′′ * isterminated(tc, h, tree′, treeleft, treeright)
         end
         return tree′, sampler′, termination′
     end
 end
 
 function transition(
     rng::AbstractRNG,
-    τ::NUTS{S,C,I,F},
     h::Hamiltonian,
+    τ::Trajectory{I, C},
+    ::Type{TS},
     z0::PhasePoint,
-) where {I<:AbstractIntegrator,F<:AbstractFloat,S<:AbstractTrajectorySampler,C<:AbstractTerminationCriterion}
+) where {I, C<:DynamicTerminationCriterion, TS}
```

Comment:
Why removing these type constraints? 
```suggestion
) where {I<:AbstractIntegrator, C<:DynamicTerminationCriterion, TS<:AbstractTrajectorySampler}
```